### PR TITLE
Eager workflow dispatch

### PR DESCRIPTION
**Pull request by bergundy**
Implement eager workflow dispatch for `StartWorkflowExecution`.

I've only added a single unit test to the repo, the rest are in [this](https://github.com/temporalio/features/pull/217) PR in the `features` repo.
Added a counter metric `workflow_eager_execution` to count the number of eager execution requests per namespace + task queue.

I've added a TODO to add support for eager signal with start, I figured that can be added later.

### CODE REVIEW

File: service/history/api/create_workflow_util.go
Comment (bergundy) [inline service/history/api/create_workflow_util.go:66]: I'm thinking of reverting this change and extracting the current task with `GetInFlightWorkflowTask`.

File: service/history/api/create_workflow_util.go
Comment (bergundy) [inline service/history/api/create_workflow_util.go:289]: I might just add support for this before merging this PR.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [inline service/history/api/startworkflow/api.go:264]: I'll figure this one out

File: service/history/api/startworkflow/api.go
Comment (bergundy) [inline service/history/api/startworkflow/api.go:278]: Could use some feedback here

**Comment by bergundy** [comment]
Looks like CI is failing for lack of coverage but there's plenty of E2E coverage in the features repo.
What's the policy here? I'd rather avoid creating a bunch of unit tests since the E2E coverage gives me much more confidence.

File: service/history/api/create_workflow_util.go
Comment (yux0) [inline service/history/api/create_workflow_util.go:168]: can it do AddFirstWorkflowTaskScheduled(...) then GetNextEventID()

Comment (bergundy) [reply service/history/api/create_workflow_util.go:168]: Sure, but that's less convenient IMHO.

File: service/history/api/startworkflow/api.go
Comment (yycptt) [inline service/history/api/startworkflow/api.go:395]: if err from GetOrCreateWorkflowExecution is not nil, then releaseFn is nil. so defer should be after this check.

File: service/history/api/startworkflow/api.go
Comment (yycptt) [inline service/history/api/startworkflow/api.go:363]: Why return error here? Start workflow is still a success.

File: service/history/api/startworkflow/api.go
Comment (yycptt) [inline service/history/api/startworkflow/api.go:426]: frontend also support return raw history event and it's controlled by a feature flag I think. Probably add a TODO for that here.

File: service/history/api/startworkflow/api.go
Comment (yycptt) [inline service/history/api/startworkflow/api.go:437]: nit: `len(response.NextPageToken) == 0` covers `response.NextPageToken == nil`

File: service/history/api/startworkflow/api.go
Comment (yycptt) [inline service/history/api/startworkflow/api.go:245]: this name is so confusing... probably `createAsCurrent`?

File: service/history/api/startworkflow/api.go
Comment (yycptt) [inline service/history/api/startworkflow/api.go:220]: L218 checked the err is not nil, then this `case nil` will never be used?

and I think the entire switch case should be moved into `applyWorkflowIDReusePolicy` as it's for handling error when there's prevExecutionUpdateAction.

File: service/history/api/create_workflow_util.go
Comment (bergundy) [reply service/history/api/create_workflow_util.go:168]: This is how the Core SDK API for constructing histories looks like.
I vote to keep this change.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:38]: Sure, I'll change.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:166]: I did not think about this case, forgot about cron workflows.
Is there something else I should be looking out for?

I'll fix the cron case though intuitively it doesn't seem like there's much of a use for cron + eager.

File: service/history/api/create_workflow_util.go
Comment (bergundy) [reply service/history/api/create_workflow_util.go:114]: Oh yes, I thought about it but forgot to add to the PR.
Thanks for the reminder.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:209]: I'll rename, thanks.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:415]: I checked and a new struct is constructed when we get it from mutable state.
Should I add a comment here?

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:363]: Because we cannot return an eager task.
Caller should be notified with an error and handle this case.
I think it's clearer than returning nil task.

File: service/history/api/create_workflow_util.go
Comment (yux0) [reply service/history/api/create_workflow_util.go:168]: i thought it could be easier to handle the 0 schedule event id. As long as we handle the 0 event id, it is fine.

File: service/history/api/create_workflow_util.go
Comment (bergundy) [reply service/history/api/create_workflow_util.go:114]: I put the config flag in the api handler (will push a fix in a bit).

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:220]: I missed this when refactoring, good catch.
I'm cleaning this up a bit.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:363]: I'm on the fence whether this should result in an error or not.
In any case what the caller should do is get a handle to the workflow and wait for its completion.
With the error approach there's at least a way to let the caller know what happened.

Comment (bergundy) [reply service/history/api/startworkflow/api.go:363]: @Spikhalskiy @cretz I could use your opinion here

File: service/history/api/create_workflow_util.go
Comment (bergundy) [reply service/history/api/create_workflow_util.go]: I put the config flag in the api handler to prevent the need to handle this throughout the codebase.
PTAL.

File: config/dynamicconfig/development-sql.yaml
Comment (bergundy) [inline config/dynamicconfig/development-sql.yaml:40]: I hope this is okay to commit, not sure what the policy is.

File: service/history/api/create_workflow_util.go
Comment (bergundy) [reply service/history/api/create_workflow_util.go:104]: I'm checking this below now.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:444]: Added todo above.

File: service/history/api/startworkflow/api.go
Comment (cretz) [reply service/history/api/startworkflow/api.go:374]: Personally, I think it's clearer/safer to consider eagerness to be a request of the server, not a requirement. The server is allowed to use whatever heuristics it wants to deny the request and do a non-eager. The absence of a task I think is enough to tell the caller the server denied the request.

I understand your use case is a "require to be eager" but that use case isn't supported for activities either. If we want a require-to-be-eager to be a thing, it should be a thing on both. I could support a server-side/namespace option of "fail if eager requested but cannot be given".

Comment (bergundy) [reply service/history/api/startworkflow/api.go:374]: Just to clarify, the question here is just whether or not to respond with an error.
The behavior is going to be the same server-side.

The client may choose to use the run ID given in the error to connect to the existing workflow in both cases.
The error path has some more information that the client might find useful.

Comment (cretz) [reply service/history/api/startworkflow/api.go:374]: Hrmm...k.

So if I call this start, may a workflow start because of it? If so, IMO that should never error even if something post-start can't happen. To me success means "workflow started because of this request" and failure means "workflow did not start because of this request". Unsure if that's related. Sorry, not familiar w/ details here so I don't have a big opinion.

Comment (bergundy) [reply service/history/api/startworkflow/api.go:374]: If a client gets this error it means that a retried request came in too late and the workflow task cannot be dispatched eagerly (likely due to it already being dispatched via the standard, matching based, path).

The only thing an error gives you over the alternative where the server responds successfully but omits the inline task is that additional piece of information.

Comment (cretz) [reply service/history/api/startworkflow/api.go:374]: Simply - if success means workflow started by this request and failure means workflow not started by request, works for me. If failure can mean workflow still started by this request, that's confusing to me.

Comment (bergundy) [reply service/history/api/startworkflow/api.go:374]: Here the workflow was started from a previous attempt of the same request. So it seems like you're saying it should not be an error.

I tend to agree but I will add a log to avoid losing some of this information.

**Review by yycptt — APPROVED** [review]

File: service/history/api/create_workflow_util.go
Comment (yux0) [inline service/history/api/create_workflow_util.go:115]: nit: for readability use newMutableState.HasPendingWorkflowTask()?

File: config/dynamicconfig/development-sql.yaml
Comment (bergundy) [reply config/dynamicconfig/development-sql.yaml:40]: Sure

File: service/history/api/create_workflow_util.go
Comment (bergundy) [reply service/history/api/create_workflow_util.go:115]: I'm not sure I find it more readable but I'm fine with making this change

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:117]: Thanks, I will add it

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:182]: I think it's clearer than the previous name but I'll change to clarify.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:212]: Yes, thanks, I fixed the comment, originally this method only handled retried requests and I forgot to update the docstring when that changed.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:374]: Ended up returning successful result and recording a metric explaining why eager execution was denied.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:473]: Oh true, so we don't need this function at all, I'll just save the first page of events

Comment (bergundy) [reply service/history/api/startworkflow/api.go:473]: Hmm.. I can't do that because I need the batches elsewhere

Comment (bergundy) [reply service/history/api/startworkflow/api.go:473]: I'd rather keep the code as is rather than panic.

Comment (yycptt) [reply service/history/api/startworkflow/api.go:497]: There're panic wrapper for all api logic, those panic won't actually bring down the service. 
But, I don't have strong opinion on it, so your call.

File: service/history/api/startworkflow/api.go
Comment (yycptt) [inline service/history/api/startworkflow/api.go:138]: `reason` is not used in ReasonTag.

File: service/history/api/startworkflow/api.go
Comment (MichaelSnowden) [inline service/history/api/startworkflow/api.go:143]: This is another place where using a parsed type instead of the raw StartRequest is better because it prevents Law of Demeter violations here and elsewhere

File: service/history/api/startworkflow/api.go
Comment (MichaelSnowden) [inline service/history/api/startworkflow/api.go:172]: Please use Errors.As

File: service/history/api/startworkflow/api.go
Comment (MichaelSnowden) [inline service/history/api/startworkflow/api.go:177]: What happens if we crash before reaching this line?

**Review by MichaelSnowden — COMMENTED** [review]
Could we abstract an `ExecutionStrategy` interface here in a separate PR to keep things SOLID and make the review easier? The code looks good to me, but it's hard to review with all of the modifications to the existing code.

File: service/history/handler.go
Comment (alexshtin) [inline service/history/handler.go:536]: This error check needs to be inside `if` block.

**Review by alexshtin — COMMENTED** [review]
WT state machine changes are not in conflict with workflow update changes.

File: common/metrics/tags.go
Comment (bergundy) [reply common/metrics/tags.go:286]: I'm not sure I follow, do you mean defining a type like:
```
// ReasonString is just a string but used to remind anyone using ReasonTag to limit the cardinality of the possible reasons.
type ReasonString string
```

If that's the case, I see little benefit to that over documenting the ask to limit the cardinality of values.
I already have a custom string enum type where this is used.

File: service/history/api/create_workflow_util.go
Comment (bergundy) [reply service/history/api/create_workflow_util.go:102]: We don't, we count on this request to be validated before this method is called.

File: service/history/handler.go
Comment (bergundy) [reply service/history/handler.go:536]: It probably doesn't matter much but you're right.

File: service/history/workflow/mutable_state.go
Comment (bergundy) [reply service/history/workflow/mutable_state.go:112]: The interface already has way too many methods IMHO.

File: service/history/workflow/mutable_state_impl.go
Comment (bergundy) [reply service/history/workflow/mutable_state_impl.go:1722]: 👍

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:68]: SGTM

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:107]: Is this something that is commonly don't in Go?
If this was JS or a functional language or a language that has the concept of immutable data I would totally clone the request.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:172]: I copied this from the original implementation but yes, As is better here.

File: service/history/api/startworkflow/api.go
Comment (bergundy) [reply service/history/api/startworkflow/api.go:177]: I don't understand the question

File: common/metrics/tags.go
Comment (bergundy) [reply common/metrics/tags.go]: I ended up adding this.

File: service/history/api/startworkflow/api.go
Comment (MichaelSnowden) [reply service/history/api/startworkflow/api.go]: For this comment:
```
	// The history and mutable state we generated above should be deleted by a background process.
```

Is `handleConflict` the method that deletes them?

File: service/history/api/startworkflow/api.go
Comment (MichaelSnowden) [reply service/history/api/startworkflow/api.go:108]: I'm thinking of the same pattern we use for tasks here: https://github.com/temporalio/temporal/blob/66db3aebeead81869dcc864e0f174063b167ee10/common/persistence/serialization/task_serializer.go

Basically taking the proto and parsing it into a plain Go object with the fields already validated and parsed into more structured types.

Comment (bergundy) [reply service/history/api/startworkflow/api.go:108]: I conceptually agree here but since the raw protos are used in the internal APIs, parsing does not make sense in this case.

**Pull request merged**
